feat: add active peer probing and a cached addr book

2color opened the pull request:
## What

This is an attempt to fix #16 by implementing #53. 

Also fixes https://github.com/ipfs/someguy/issues/25
## How

- New Cached Address Book
  - Wraps [`memoryAddrBook`](https://github.com/libp2p/go-libp2p/blob/6bc70bf4a9cf8cfe38d158e47d11fceed175d921/p2p/host/peerstore/pstoremem/addr_book.go/#L187) which
- New Cached Router and enriches results with cached addresses when records have no addresses.
  - Implement a custom iterator for FindProviders that looks up cache, returns result with addrs if there's a cache HIT, or dispatches a FindPeer if there's a cache miss, which it then returns to the user once a result is rerturned
- New background goroutine
  - Subscribes to identify and connectedness events and updates cached address book. 
  - Runs a probe against all peers that meet probe criteria:
    - Not currently connected
    - Haven't been probed in the last threshold (1 hour)
## New magic numbers

We have to start with some default. This PR introduces some magic numbers which will likely change as we get some operational data:

https://github.com/ipfs/someguy/blob/19b15aabc2fc6be3335ddc0e7717512f522fe42b/server_addr_book.go#L21-L37


## Open questions
- The only peers we probe and cache are peers that we've successfully run identify with. Peers returned without multiaddrs from `FindProviders` for which we have no cached multiaddrs remain unresolved. 
  - Should we try to call FindPeer inside the iterator so they can be resolved? This can blocking the streaming of othe providers in the iterator.
  - Another way might be to subscribe to kad-dht query events (not 100% sure if this is possible) and add to probe loop
- Should we probe the last connected addr or all addresses we have for a Peer?
- When should we augment results with cached addresses? Currently, it's done only when there are no results in the FindProviders from kad-dht. The presumption there is that if you get the results from `FindProviders` have multiaddrs  for a peer, it's up to date.
- How do we prevent excessive memory consumption by the cached address book? The memory address book already has built in limits and clean up. However, the `peers` map doesn't. **temp solution: I've added some instrumentation for this**

2color committed:
feat: add cached peer book with higher ttls

2color committed:
feat: initial implementation of active peer probing

2color committed:
feat: use the cached router

2color committed:
chore: go mod tidy

2color committed:
feat: log probe duration

2color committed:
chore: log in probe loop

2color committed:
fix: update peer state if doesn't exist

2color committed:
fix: add addresses to cached address book

2color committed:
fix: wrap with cached router only if available

2color committed:
feat: make everything a little bit better

2color committed:
chore: small refinements

2color committed:
test: add test for cached addr book

2color committed:
chore: rename files

2color committed:
feat: add options to cached addr book

fix test by allowing private ips

2color committed:
feat: add instrumentation

2color committed:
fix: thread safety

2color committed:
docs: update changelog

2color committed:
fix: small fixes

2color committed:
fix: simplify cached router

lidel committed:
feat(metric): cached_router_peer_addr_lookups

this adds metric for evaluating all addr lookups
someguy_cached_router_peer_addr_lookups{cache="unused|hit|miss",origin="providers|peers"}

I've also wired up FindPeers for completeness.

lidel reviewed (commented):
Made a first pass and dropped some suggestions inline. I also pushed  with new metric (details inline).

As for Open questions, my thinking is:


> - The only peers we probe and cache are peers that we've successfully run identify with. Peers returned without multiaddrs from `FindProviders` for which we have no cached multiaddrs remain unresolved. 
>   - Should we try to call FindPeer inside the iterator so they can be resolved? This can blocking the streaming of the providers in the iterator.

Indeed, looking at `someguy_cached_router_peer_addr_lookups` shows we have cache `miss` quite often (0 addrs + cache also does not have them).

Was bit difficult to reason without some real-world input, so I've  piped root CIDs hitting our staging environment, to populate the metric:

- with CID duplicates: `ssh ubuntu@kubo-staging-us-east-02.ovh.dwebops.net tail -F /var/log/nginx/access-json.log | stdbuf -o0 jq -r .request_uri | awk -F'[/&?]' '{print $3; fflush()}' | xargs -P 100 -I {cid} curl -s -o /dev/null -w "%{http_code} %{url.path}\n" "http://127.0.0.1:8190/routing/v1/providers/{cid}"`
- only unique CIDs: `ssh ubuntu@kubo-staging-us-east-02.ovh.dwebops.net tail -F /var/log/nginx/access-json.log | stdbuf -o0 jq -r .request_uri | awk -F'[/&?]' '!seen[$3]++ {print $3; fflush()}' | xargs -P 100 -I {cid} curl -s -o /dev/null -w "%{http_code} %{url.path}\n" "http://127.0.0.1:8190/routing/v1/providers/{cid}"`

A few minutes later http://127.0.0.1:8190/debug/metrics/prometheus shows:

```
# HELP someguy_cached_router_peer_addr_lookups Number of peer addr info lookups per origin and cache state
# TYPE someguy_cached_router_peer_addr_lookups counter
someguy_cached_router_peer_addr_lookups{cache="hit",origin="providers"} 1323
someguy_cached_router_peer_addr_lookups{cache="miss",origin="providers"} 6574
someguy_cached_router_peer_addr_lookups{cache="unused",origin="providers"} 7686
```

So yes, finding a way of decreasing `miss` feels useful, given how high it is.

Two ideas: 

- Lazy/easy: avoid blocking iterator by adding peers with cache misses to some queue, and then processing them asynchronously at some safe rate, populating cache in best-effort fashion. May not help first query, but all subsequent ones, over time, will get increased cache `hit`
- Implement custom iterator: if peer hits cache miss, we dont return the peer, but silently moves to the next item, and puts current one at the side queue which is processed async calling findPeer. once the iterator hits the last item, we go back to items on the side queue. This way we don't slow down results with addrs, and we can wait and stream ones at the end without impacting perf of  fast ones. 

> - Should we probe the last connected addr or all addresses we have for a Peer?

See comment inline, iiuc `host.Connect` effectively probes all of known addrs, until success.
Probably good enough for now. If we need per-addr resolution, we may need ask go-libp2p for new API.

Note that `vole libp2p identify <multiaddr>` connects to specific multiaddr because it does not run routing and [spawns a new libp2p host every time](https://github.com/ipfs-shipyard/vole/blob/b682bf36a6fb2ecd0c88791b184755174007d612/lib/identify.go#L48).

> - When should we augment results with cached addresses? Currently, it's done only when there are no results in the FindProviders from kad-dht. The presumption there is that if you get the results from `FindProviders` have multiaddrs  for a peer, it's up to date.

I think current approach of hitting cache if regular routing returns no addrs is sensible.
It also makes it  easier to reason about metrics like `someguy_cached_router_peer_addr_lookups{origin,cache}`

> - How do we prevent excessive memory consumption by the cached address book? The memory address book already has built in limits and clean up. However, the `peers` map doesn't. **temp solution: I've added some instrumentation for this**

Cap at TTL of 48h?

2color commented on server_cached_router.go:
```diff
@@ -65,22 +84,54 @@ func (r cachedRouter) FindProviders(ctx context.Context, key cid.Cid, limit int)
 }
 
 func (r cachedRouter) FindPeers(ctx context.Context, pid peer.ID, limit int) (iter.ResultIter[*types.PeerRecord], error) {
```
Btw, initially I had implemented the cache for `FindPeers` only to remove it later. My reasoning is that if you fail to `FindPeer`, as part of a kad-dht look (which will ask at least 20 peers), you probably don't want to return it from cache.

Do you still see value in returning addresses from cache here? At the very least we could maybe check the last time we connected to them in `withAddrsFromCache`/`GetCachedAddrs`.

2color commented on cached_addr_book.go:
```diff
@@ -0,0 +1,275 @@
+	// Connected peers don't expire until they disconnect
+	ConnectedAddrTTL = math.MaxInt64
+
+	// How long to wait since last connection before probing a peer again
+	PeerProbeThreshold = time.Hour
```
If you search for references to `peerstore.AddressTTL` you'll see it isn't used by anything (both in go-libp2p and go-libp2p-kad-dht) so maybe we should remove it?

There are two relevant address TTLs:
- [RecentlyConnectedTTL](https://github.com/libp2p/go-libp2p/blob/8423de3a64f17f6bec18bf57b472e5a3615883db/core/peerstore/peerstore.go#L29-L31) which is 15 minutes 
- [ProviderAddrTTL](https://github.com/libp2p/go-libp2p-kad-dht/blob/25e4aff78feaf7a83c0ebde80b538521210292b6/amino/defaults.go/#L49) which is 24 hours and is only for addresses associated with a provider record you are storing (so much less prevalent)


If we want to probe after the go-libp2p TTL expires, that would have to be 15 minutes for most addresses, but probing that frequently would mean we probe almost every peer every `probePeers` run, since that's how often the probe runs. Once we histogram data from production on how long `probePeers` takes, we can adjust.

2color committed:
Apply suggestions from code review

Co-authored-by: Marcin Rataj <lidel@lidel.org>

2color committed:
Update CHANGELOG.md

Co-authored-by: Marcin Rataj <lidel@lidel.org>

2color commented on server_cached_router.go:
```diff
@@ -0,0 +1,137 @@
+var (
+	_ server.ContentRouter = cachedRouter{}
+
+	// peerAddrLookups allows us reason if/how effective peer addr cache is
+	peerAddrLookups = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name:      "peer_addr_lookups",
+		Subsystem: "cached_router",
+		Namespace: "someguy",
+		Help:      "Number of peer addr info lookups per origin and cache state",
+	},
+		[]string{addrCacheStateLabel, addrQueryOriginLabel},
```
Nice

2color committed:
chore: use service name for namespace

2color committed:
fix: type errors and missing imports

2color committed:
feat: add queue probe

2color committed:
Revert "feat: add queue probe"

This reverts commit 75f1bf270a9462bbe9a146c5f8887c12a8866810.

2color committed:
chore: simplify composite literal

2color committed:
fix: implement custom cache fallback iterator

2color committed:
fix: add cancel and simplify

2color committed:
fix: move select to Val function

2color committed:
fix: concurrency bug from the ongoingLookups

2color committed:
chore: clean up comments

2color committed:
fix: add lint ignores

2color committed:
docs: update changelog

2color committed:
fix: increase bucket sizes for probe duration

2color committed:
chore: remove unused peer state fields

save some memory

2color committed:
feat: enable caching for FindPeer in cached router

2color committed:
fix: handle peer not found case

lidel commented on cached_addr_book.go:
```diff
@@ -0,0 +1,275 @@
+					logger.Debug("Caching signed peer record")
+					cab, ok := peerstore.GetCertifiedAddrBook(cab.addrBook)
+					if ok {
+						ttl := RecentlyConnectedAddrTTL
+						if host.Network().Connectedness(ev.Peer) == network.Connected || host.Network().Connectedness(ev.Peer) == network.Limited {
+							ttl = ConnectedAddrTTL
+						}
```
Fair enough, but then maybe extract all those checks into a func, like in kad-dht?

2color replied:
Fixed in 22aacd7b0a5478af7e198cea866c4b9960e0a30d

lidel reviewed (commented):
Had to review at a later hours, so ymmv, but hopefully useful feedback:
- main ask: avoid dispatching peer lookup for peers that are offline (more inline)
- re: cleanup for the cachedAddrBook.peers
  -  we increment `peerState.returnCount` every time we read form cache, we could also track timestamp and GC peers that were not read from cache for  longer than our probe interval or something
  - but, implementing own GC and cache is usually last thing you want to do, see if you could reuse [`TwoQueueCache`](https://github.com/hashicorp/golang-lru/blob/main/2q.go) from https://github.com/hashicorp/golang-lru – use defaults, just set its max size to something like 10k (`someguy_cached_addr_book_peer_state_size` seems to be aroundthat) and it will take care of evicting least useful entries that fall beyond frecency threshold.
- re: TestBackground 
  - i've run out of time to look, but rule of thumb is to avoid channels for time-dependent code / sleep 
  - maybe pass `ctx, cancel := context.WithTimeout` to `go cab.background(ctx`, remove other goroutines, and wait in a single loop at the end, something along these lines:
    ```go
    ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
    // ...
    go cab.background(ctx, h)
    // ...
    for {
            select {
            case <-ctx.Done():
                t.Fatal("timeout waiting for peer to be added to peer state")
            default:
                cab.mu.RLock()
                _, exists := cab.peers[testPeer]
                cab.mu.RUnlock()
                if exists {
                    return
                }
                time.Sleep(10 * time.Millisecond)
            }
        }
    ```

- ps. I realized the piped traffic will include duplicate CIDs due to requests being for multiple files under same root. If you want to pipe only unique CIDs, use  `| awk -F'[/&?]' '!seen[$3]++ {print $3; fflush()}' |`
  - Using unique CIDs does not change much, hit/miss/unused ratios are similar, just give bit more confidence:
    ```
    # HELP someguy_cached_addr_book_peer_state_size Number of peers object currently in the peer state
    # TYPE someguy_cached_addr_book_peer_state_size gauge
    someguy_cached_addr_book_peer_state_size 9738

    # HELP someguy_cached_router_peer_addr_lookups Number of peer addr info lookups per origin and cache state
    # TYPE someguy_cached_router_peer_addr_lookups counter
    someguy_cached_router_peer_addr_lookups{cache="hit",origin="peers"} 1
    someguy_cached_router_peer_addr_lookups{cache="hit",origin="providers"} 3544
    someguy_cached_router_peer_addr_lookups{cache="miss",origin="peers"} 1173
    someguy_cached_router_peer_addr_lookups{cache="miss",origin="providers"} 1352
    someguy_cached_router_peer_addr_lookups{cache="unused",origin="peers"} 178
    someguy_cached_router_peer_addr_lookups{cache="unused",origin="providers"} 7402
    ```

2color commented on server_cached_router.go:
```diff
@@ -0,0 +1,244 @@
+func (it *cacheFallbackIter) Close() error {
+	it.cancel()
+	go func() {
+		for it.ongoingLookups.Load() > 0 {
+			time.Sleep(time.Millisecond * 10)
+		}
```
Yeah the idea was to ensure we only close the channel after dispatched goroutines write to the channel. 

I'll increase to 100ms

2color replied:
fixed in ecd075795fe6cd337e9b34f9478173207291c248

2color committed:
Apply suggestions from code review

Co-authored-by: Marcin Rataj <lidel@lidel.org>

2color committed:
fix: wait longer during cleanup function

2color committed:
test: remove bitswap record test

2color committed:
refactor: extract connectedness checks to a func

2color committed:
fix: set ttl for both signed and unsigned addrs

2color committed:
fix: prevent race condition

2color commented on server_cached_router.go:
```diff
@@ -0,0 +1,244 @@
+			logger.Infow("no cached addresses found in cacheFallbackIter, dispatching find peers", "peer", id)
+			// If a record has no addrs, we dispatch a lookup to find addresses
+			go it.dispatchFindPeer(*record)
+			// important to increment here since Next() may be called again synchronously
+			it.ongoingLookups.Add(1)
```
fixed in 03a4078b817bd3c401d8912ac9535dc3aa1d658f

2color replied:
fixed in 03a4078b817bd3c401d8912ac9535dc3aa1d658f

2color commented on server_cached_router.go:
```diff
@@ -0,0 +1,244 @@
+			//lint:ignore SA1019 // ignore staticcheck
+			if record, ok := val.Val.(*types.BitswapRecord); ok {
+				return handleRecord(record.ID, types.FromBitswapRecord(record))
+			}
```
Thanks for looking into this. Makes sense

2color committed:
feat: use 2q-lru cache for peer state

2q-lru tracks both frequently and recently used entries separately

2color committed:
chore: remove return count

we don't need the return count with the 2q-lru cache and the peerAddrLookups
metric

2color committed:
test: improve reliability of tests

mock the libp2p host and use a real event bus

2color committed:
fix: record failed connections

2color committed:
feat: add exponential backoff for probes/peer lookups

2color committed:
fix: return peers with no addrs that wont probe

2color commented on server_cached_router.go:
```diff
@@ -0,0 +1,244 @@
+			}
+			logger.Infow("no cached addresses found in cacheFallbackIter, dispatching find peers", "peer", id)
+			// If a record has no addrs, we dispatch a lookup to find addresses
+			go it.dispatchFindPeer(*record)
```
I've added backoff in b536e82ff40bf1134cb53407811089e6cb957f91

2color commented:
Thanks @lidel. I've addressed all your points.

2color committed:
fix: brittle test

2color committed:
feat: add probed peers counter

2color committed:
fix: adjust probe duration metric buckets

2color commented:
I've been running this for a little while with the accelerated DHT client (similar to how production runs) 

And got these metrics:

```
# HELP someguy_cached_addr_book_peer_state_size Number of peers object currently in the peer state
# TYPE someguy_cached_addr_book_peer_state_size gauge
someguy_cached_addr_book_peer_state_size 9729
# HELP someguy_cached_addr_book_probe_duration_seconds Duration of peer probing operations in seconds
# TYPE someguy_cached_addr_book_probe_duration_seconds histogram
someguy_cached_addr_book_probe_duration_seconds_bucket{le="1"} 0
someguy_cached_addr_book_probe_duration_seconds_bucket{le="2"} 0
someguy_cached_addr_book_probe_duration_seconds_bucket{le="5"} 0
someguy_cached_addr_book_probe_duration_seconds_bucket{le="10"} 0
someguy_cached_addr_book_probe_duration_seconds_bucket{le="30"} 0
someguy_cached_addr_book_probe_duration_seconds_bucket{le="60"} 0
someguy_cached_addr_book_probe_duration_seconds_bucket{le="120"} 0
someguy_cached_addr_book_probe_duration_seconds_bucket{le="300"} 9
someguy_cached_addr_book_probe_duration_seconds_bucket{le="+Inf"} 12
someguy_cached_addr_book_probe_duration_seconds_sum 3197.8019163719996
someguy_cached_addr_book_probe_duration_seconds_count 12
```

Which reveal an average of ~5 minutes for a probe cycle. 

I therefore made the following changes to the instrumentation:
1. Add a probedPeersCounter to track how many individual peers are probed 
2. Increased the probe_duration bucket sizes so we have an eye if these get too long as the size of the address book gets large (as it happens with the accelerated DHT client)

sukunrt commented on server_cached_router.go:
```diff
@@ -0,0 +1,247 @@
+				it.ongoingLookups.Add(1) // important to increment before dispatchFindPeer
+				// If a record has no addrs, we dispatch a lookup to find addresses
+				go it.dispatchFindPeer(*record)
+
+				return it.Next() // Recursively call Next() to either read from sourceIter or wait for lookup result
```
Will this exit the loop if `it.Next()` returns false? Do you need to wait for the find peer call dispatched just before to end?

sukunrt replied:
This is probably my unfamiliarity with the construct. Please ignore if it doesn't make sense.

2color replied:
Great question. The recursive `it.Next()` call will block on ongoing lookups (if the underlying iterator is done but we have an ongoing dispatched find peer), thereby ensuring that we return the value based on whatever the result of the dispatched lookup.

sukunrt replied:
Ah, my bad. For some reason I read that as source.Next.

sukunrt commented on cached_addr_book.go:
```diff
@@ -0,0 +1,307 @@
+// Returns true if we should probe a peer (either by dialing known addresses or by dispatching a FindPeer)
+// based on the last failed connection time and connection failures
+func (cab *cachedAddrBook) ShouldProbePeer(p peer.ID) bool {
+	pState, exists := cab.peerCache.Get(p)
+	if !exists {
+		return true // default to probing if the peer is not in the cache
+	}
```
Why is this the default? 
An expired peer is not recent or frequent, should you drop it?

2color replied:
Good question. If it reaches this code block, the peer is in the address book (which isn't perfectly aligned with the `peerCache`), which means it was encountered and was online in the last 48 hours. The circumstances under which this scenario happens are unlikely (because both address book and cache are populated in the same place), and if they happen it's when both are almost full. 

If this probe fails, the backoff will kick in, but seems useful to at least try if we assume that there might be newly offline other peers in the address book.

sukunrt replied:
makes sense.

sukunrt commented on cached_addr_book.go:
```diff
@@ -0,0 +1,307 @@
+	if pState.connectFailures > 0 {
+		// Calculate backoff only if we have failures
+		// this is effectively 2^(connectFailures - 1) * PeerProbeThreshold
+		// A single failure results in a 1 hour backoff
+		backoffDuration = PeerProbeThreshold * time.Duration(1<<(pState.connectFailures-1))
```
NIT: Should this have an upper bound?

2color replied:
Maybe. eventually it'll either get cleaned from the cache if it reaches max capacity or will not even reach here if 48 hours pass (and the peer TTL expires in the address book) unless the peer is encountered again via a FindPeer/FindProviders call to the service.

sukunrt replied:
I'd set the BackoffDuration to `min(backoffDuration, MaxBackoffDuration)`. This would mean eventually we will probe a peer once a day(at least). The current method is also fine though. I don't have a strong opinion here.

sukunrt commented on cached_addr_book.go:
```diff
@@ -0,0 +1,307 @@
+				logger.Debug("Skipping peer probe, still running")
+				continue
+			}
+			logger.Debug("Starting to probe peers")
+			go cab.probePeers(ctx, host)
+		}
+	}
+}
+
+// Loops over all peers with addresses and probes them if they haven't been probed recently
+func (cab *cachedAddrBook) probePeers(ctx context.Context, host host.Host) {
+	cab.isProbing.Store(true)
+	defer cab.isProbing.Store(false)
```
You're unlikely to hit this in production. 

The spawned goroutine may run arbitrarily later. So while the goroutine is waiting to be run, the ticker might go off again and start another goroutine, as the isProbing variable is updated in the goroutine.

2color replied:
By that logic, can I just move the `Store` to the `background` to avoid the anonymous function:

```
case <-probeTicker.C:
			if cab.isProbing.Load() {
				logger.Debug("Skipping peer probe, still running")
				continue
			}
			logger.Debug("Starting to probe peers")
			cab.isProbing.Store(true)
			go cab.probePeers(ctx, host)
		}
	}
}

// Loops over all peers with addresses and probes them if they haven't been probed recently
func (cab *cachedAddrBook) probePeers(ctx context.Context, host host.Host) {
	defer cab.isProbing.Store(false)
```

2color committed:
fix: prevent race conditions

sukunrt commented on cached_addr_book.go:
```diff
@@ -0,0 +1,307 @@
+	// The worst case is a peer behind a relay, so we use the relay connect timeout.
+	ConnectTimeout = relay.ConnectTimeout
+
+	// How many peers to cache in the peer state cache
+	// 100_000 is also the default number of signed peer records cached by the memory address book.
+	PeerCacheSize = 100_000
```
You can make this much larger(1M?). Compared to the peerstore the info stored here is much smaller. Also this has negative results for failed queries.

2color committed:
feat: increase cache size and add max backoff

lidel commented on server_cached_router.go:
```diff
@@ -0,0 +1,247 @@
+	peersIt, err := it.router.FindPeers(it.ctx, *record.ID, 1)
+
+	if err != nil {
+		it.findPeersResult <- record // pass back the record with no addrs
```
@aschmahmann noted that if we made probe and fail to find addrss for that PeerID, there is no point returning that PeerID without addrs, because client will try to do the same lookup (potentially hitting /routing/v1/peers too), and fail again. 

So let's skip PeerIDs that have no addrs and fail FindPeers here.

2color replied:
Fixed in https://github.com/ipfs/someguy/pull/90/commits/c5abeec8c18055d9c3c871db2f387ac819f38e27

2color replied:
@lidel Should we also omit cached addresses from `FindPeer` requests? Currently, if the libp2p `FindPeer` fails, we try returning from the cache. But maybe that's a bad idea? 

https://github.com/ipfs/someguy/blob/c5abeec8c18055d9c3c871db2f387ac819f38e27/server_cached_router.go#L70-L99

2color replied:
As discussed in person with @lidel. We will avoid looking up the cache in FindPeer queries, since the FindPeer should always return whether a peer is currently dialable. 

We will only record connection failures for probing purposes.

2color committed:
fix: omit providers whose peer cannot be found

2color committed:
chore: remove unused function

2color committed:
deps: upgrade go-libp2p

2color committed:
fix: avoid using the cache in FindPeers

2color committed:
fix: do not return cached results for FindPeers

2color committed:
refactor: small optimisation

2color committed:
chore: re-add comment

lidel commented on cached_addr_book.go:
```diff
@@ -0,0 +1,311 @@
+		case ev := <-sub.Out():
+			switch ev := ev.(type) {
+			case event.EvtPeerIdentificationCompleted:
+				pState, exists := cab.peerCache.Get(ev.Peer)
```
nit: use `Peek` instead of `Get` when we don't have to bump 2Q frecency (e.g. here we call `Add` few lines lower, so no need to use `Get`)
```suggestion
				pState, exists := cab.peerCache.Peek(ev.Peer)
```

lidel commented on cached_addr_book.go:
```diff
@@ -0,0 +1,311 @@
+// Returns true if we should probe a peer (either by dialing known addresses or by dispatching a FindPeer)
+// based on the last failed connection time and connection failures
+func (cab *cachedAddrBook) ShouldProbePeer(p peer.ID) bool {
+	pState, exists := cab.peerCache.Get(p)
```
```suggestion
	pState, exists := cab.peerCache.Peek(p)
```

lidel commented on server_cached_router.go:
```diff
@@ -0,0 +1,230 @@
+		val := it.sourceIter.Val()
+		handleRecord := func(id *peer.ID, record *types.PeerRecord) bool {
+			record.Addrs = it.router.withAddrsFromCache(addrQueryOriginProviders, *id, record.Addrs)
+			if record.Addrs != nil {
```
nit: more idiomatic go is to avoid nil checks, length ones are less error prone during refactors (len will work with nil and empty slice)
```suggestion
			if len(record.Addrs) != 0 {
```

lidel commented on server_cached_router.go:
```diff
@@ -0,0 +1,230 @@
+				// If a record has no addrs, we dispatch a lookup to find addresses
+				go it.dispatchFindPeer(*record)
+
+				return it.Next() // Recursively call Next() to either read from sourceIter or wait for lookup result
+			}
+
+			// If we're not going to probe, return the record with no addrs
+			// TODO: should we even return these if the peer is likely unreachable?
+			it.current = iter.Result[types.Record]{Val: record}
+			return true
```
:warning: if we did active probing before, we should skip those empty peers for sure:

```suggestion
			}
			return it.Next() // Recursively call Next() to either read from sourceIter or wait for lookup result
```


Rationale:

(in most cases they will cause clients to do useless peer lookups that will fail as well)

Currently (with those empty results) we have bizzare behavior where initial  lookup returns only online peers, but subsequent lookup starts returning offline peers that failed initial probe.

Practical example: lets find providers for /ipns/protocol.ai:

```console
$ curl "http://127.0.0.1:8190/routing/v1/providers/QmbKiad69C9Ucy3enEyKpmeyiKk2NPi628dDs3PanTKR1P" -H 'Accept: application/x-ndjson' -s  
{"Addrs":["/ip4/69.203.71.93/udp/64283/quic-v1/p2p/12D3KooWNMURc4AuvjC1vYARFTSKmMJBB1pjyr5zymBL15E96rqA/p2p-circuit","/ip4/95.216.28.167/udp/4001/quic/p2p/12D3KooWPDBio8GH9svF7oS1mw849esxLs3gr5YViUGqzi2zwef9/p2p-circuit","/ip4/95.216.28.167/tcp/4001/p2p/12D3KooWPDBio8GH9svF7oS1mw849esxLs3gr5YViUGqzi2zwef9/p2p-circuit","/ip4/69.203.71.93/udp/64283/quic-v1/webtransport/certhash/uEiAZ3F5UJwcCIr3UPr_sPuWQj3v4HAasik_H_dLt0bzZTQ/certhash/uEiAPUcxZdKwDMsPCOs8qe4rmvLJRdNUVH6Mhi2nSy6EniQ/p2p/12D3KooWNMURc4AuvjC1vYARFTSKmMJBB1pjyr5zymBL15E96rqA/p2p-circuit","/ip4/69.203.71.93/tcp/64283/p2p/12D3KooWNMURc4AuvjC1vYARFTSKmMJBB1pjyr5zymBL15E96rqA/p2p-circuit"],"ID":"12D3KooWSXS5KPLpPZ4jfCfek5SrCW2EeGFfkQr3fqb3UAikfX58","Schema":"peer"}
{"Addrs":["/ip4/216.238.81.231/udp/4001/quic-v1/p2p/12D3KooWRo1uYUSm9wXn2LeBpZyQrU3KYeq2yzkKRSGc8FYjDXpb/p2p-circuit","/ip4/216.238.81.231/tcp/4001/p2p/12D3KooWRo1uYUSm9wXn2LeBpZyQrU3KYeq2yzkKRSGc8FYjDXpb/p2p-circuit","/ip4/216.238.81.231/udp/4001/quic/p2p/12D3KooWRo1uYUSm9wXn2LeBpZyQrU3KYeq2yzkKRSGc8FYjDXpb/p2p-circuit","/ip4/66.29.147.177/tcp/4001/p2p/12D3KooWJNX7YDRBkh1m2vssumMHoP7PEn2gQ7WVs5s3vYxGfxxb/p2p-circuit","/ip4/66.29.147.177/udp/4001/quic-v1/webtransport/certhash/uEiBjZkayyaLxp88App6om86GYU2q-TB8I50gowq5MCCPqQ/certhash/uEiCwAXs3XknzDnLuXHkvwiJn-StgfExwnGPR-7pA5zsrmg/p2p/12D3KooWJNX7YDRBkh1m2vssumMHoP7PEn2gQ7WVs5s3vYxGfxxb/p2p-circuit","/ip4/66.29.147.177/udp/4001/quic-v1/p2p/12D3KooWJNX7YDRBkh1m2vssumMHoP7PEn2gQ7WVs5s3vYxGfxxb/p2p-circuit"],"ID":"12D3KooWEp8H3jDYw1Cn3hL432UX71Wv4UKmTXHP4mmL597GTmVn","Schema":"peer"}
{"Addrs":["/ip4/145.40.89.101/tcp/4001","/ip6/2604:1380:45f1:d800::1/tcp/4001"],"ID":"12D3KooWHVXoJnv2ifmr9K6LWwJPXxkfvzZRHzjiTZMvybeTnwPy","Schema":"peer"}
{"Addrs":["/ip4/145.40.89.101/udp/4001/quic-v1/webtransport/certhash/uEiBO8uH1QnET8rmwLLXlR5rocnu8gjZ1TZDQdMdp_yOowA/certhash/uEiA7PuBdt9syTOi4s3zSDE1XD2kBGN__66nSYaszeH3xRA","/ip6/2604:1380:45f1:d800::1/tcp/4002/tls/sni/2604-1380-45f1-d800--1.k51qzi5uqu5dj0wvrbb8keygfyxe2v0fi1qbqz4pl3zzozle7oaqhf97mqazo4.libp2p.direct/ws","/ip4/145.40.89.101/tcp/4002/tls/sni/145-40-89-101.k51qzi5uqu5dj0wvrbb8keygfyxe2v0fi1qbqz4pl3zzozle7oaqhf97mqazo4.libp2p.direct/ws","/ip4/145.40.89.101/udp/4001/quic-v1","/ip6/2604:1380:45f1:d800::1/tcp/4001","/ip6/2604:1380:45f1:d800::1/udp/4001/webrtc-direct/certhash/uEiAE_vuhjpPz5yq87PvAUAWca2yJxiuByzd_wRKiqcJGlw","/ip6/2604:1380:45f1:d800::1/udp/4001/quic-v1/webtransport/certhash/uEiBO8uH1QnET8rmwLLXlR5rocnu8gjZ1TZDQdMdp_yOowA/certhash/uEiA7PuBdt9syTOi4s3zSDE1XD2kBGN__66nSYaszeH3xRA","/ip4/145.40.89.101/udp/4001/webrtc-direct/certhash/uEiAE_vuhjpPz5yq87PvAUAWca2yJxiuByzd_wRKiqcJGlw","/ip4/145.40.89.101/tcp/4001","/ip6/2604:1380:45f1:d800::1/udp/4001/quic-v1"],"ID":"12D3KooWHVXoJnv2ifmr9K6LWwJPXxkfvzZRHzjiTZMvybeTnwPy","Schema":"peer"}
```

Then calling second time returns empty ones, likely  random people who visited /ipns/protocol.ai with their desktop node that is currently offline:

```console
$ curl "http://127.0.0.1:8190/routing/v1/providers/QmbKiad69C9Ucy3enEyKpmeyiKk2NPi628dDs3PanTKR1P" -H 'Accept: application/x-ndjson' -s
{"ID":"12D3KooWAMhEcutSZRHP9BiSKa8Tsz3cFefKBSezhJRqyb62rLFg","Schema":"peer"}
{"ID":"12D3KooWCUW1rNfUtyJCATxCh1yQhVdwMUJYywLYWk5X9FgTxrPT","Schema":"peer"}
{"ID":"12D3KooWCoqS7A7MBy7pbAhJbPNwXytwYkBC12A7h3frUe3Tip6p","Schema":"peer"}
{"ID":"12D3KooWGqzJNT8RSg5YPpProaeYqXQKLg8k7grChnyL4dMp1sED","Schema":"peer"}
{"ID":"12D3KooWJiVts67y1rbQbKJMuKLzLBjnMu7xLB7o6BxXMBNp1pNG","Schema":"peer"}
{"ID":"12D3KooWPJaWvf5W8Zhss97CvDzs9LpNuewHFvUVZTFFkyAcHcM2","Schema":"peer"}
{"ID":"12D3KooWQCbo62Cs8uXPJPdrW3S7DQCbSABezR3mMud4uSzetMeT","Schema":"peer"}
{"ID":"12D3KooWQa3hz3LsGMVfsUSwJUFurekidxAw8QaNyrqxXHr3iE2m","Schema":"peer"}
{"ID":"12D3KooWDot6SE3cVqAPe6Rwk1zunSrUTg9Z1RmB5YHgouEnQ6Zx","Schema":"peer"}
{"ID":"12D3KooWE3CpyihkLaSGAZEyEJ9sWijByTZ4HHqiC4w3thqufoSs","Schema":"peer"}
{"ID":"12D3KooW9x8nmeaP2i4SZD27Q6Ew6zNrffBnbLnMqCmzzW9PtZtF","Schema":"peer"}
{"ID":"12D3KooW9uKfmrqDXmiKAcGw3Q5csKjMPBcg7aBgXPToe8GpYRun","Schema":"peer"}
{"ID":"12D3KooW9yycnejq1rRXc16bUGBQWDeLPXd81UG4LwRPg621uYVX","Schema":"peer"}
{"ID":"12D3KooWA2viJLxkVNrs8v1FW59q8qopR99GueJ11YYzSpT9wVZo","Schema":"peer"}
{"ID":"12D3KooWARy1rzqfHU1w3aHgXxpk1WpmAcRUVqc31xcwr5ZmYhu3","Schema":"peer"}
{"ID":"12D3KooWAX7yomznUC1HXAh2MuXkAxCSGkBA3FsUHhS53PvSc7A7","Schema":"peer"}
{"ID":"12D3KooWCPgSBDH637g5sAxJRLuSngefzMGUeNJEtJkjabk3qmB1","Schema":"peer"}
{"ID":"12D3KooWCq9R7DmRbxgwHYc6FwzmJL6tZMK1qMhTVqWbmf8GbSvZ","Schema":"peer"}
{"ID":"12D3KooWEc9p8exRiDzeNofnxASWhk7r9BEUgY8rPYdHJ8MGmE8u","Schema":"peer"}
{"ID":"12D3KooWFnbMPGB2u26Z8Kv2ZTxoHZaucSUczJEH9xXogjnnSxsV","Schema":"peer"}
{"ID":"12D3KooWFogZTo5LTsveft62VRGjVCFkhPGWrSyvdPy5xkuVrWPZ","Schema":"peer"}
{"ID":"12D3KooWFqq1yzb65kWTp9oMXyYS9HuVmCYxj8STCf9TiVpJXogr","Schema":"peer"}
{"ID":"12D3KooWGjcvV93ThR9E1HgsEDG7yzbKbjHBj6iZusDpkg9yNR44","Schema":"peer"}
{"ID":"12D3KooWH5xqQ39okv5PpxwdPWCNyWgLqJ6vv58eyo5auqDNJTy2","Schema":"peer"}
{"ID":"12D3KooWNDXvi4vQn8xBtQXCcMiUA65zY9yfJF7FQehjtKA2Yv4d","Schema":"peer"}
{"Addrs":["/ip4/127.0.0.1/tcp/4001","/ip6/2604:1380:45f1:d800::1/udp/4001/quic-v1/webtransport/certhash/uEiBO8uH1QnET8rmwLLXlR5rocnu8gjZ1TZDQdMdp_yOowA/certhash/uEiA7PuBdt9syTOi4s3zSDE1XD2kBGN__66nSYaszeH3xRA","/ip6/::1/udp/4001/webrtc-direct/certhash/uEiAE_vuhjpPz5yq87PvAUAWca2yJxiuByzd_wRKiqcJGlw","/ip4/145.40.89.101/udp/4001/webrtc-direct/certhash/uEiAE_vuhjpPz5yq87PvAUAWca2yJxiuByzd_wRKiqcJGlw","/ip4/145.40.89.101/udp/4001/quic-v1","/ip4/127.0.0.1/udp/4001/quic-v1/webtransport/certhash/uEiBO8uH1QnET8rmwLLXlR5rocnu8gjZ1TZDQdMdp_yOowA/certhash/uEiA7PuBdt9syTOi4s3zSDE1XD2kBGN__66nSYaszeH3xRA","/ip4/145.40.89.101/tcp/4001","/ip4/145.40.89.101/tcp/4002/tls/sni/145-40-89-101.k51qzi5uqu5dj0wvrbb8keygfyxe2v0fi1qbqz4pl3zzozle7oaqhf97mqazo4.libp2p.direct/ws","/ip6/::1/udp/4001/quic-v1","/ip6/::1/udp/4001/quic-v1/webtransport/certhash/uEiBO8uH1QnET8rmwLLXlR5rocnu8gjZ1TZDQdMdp_yOowA/certhash/uEiA7PuBdt9syTOi4s3zSDE1XD2kBGN__66nSYaszeH3xRA","/ip6/2604:1380:45f1:d800::1/udp/4001/webrtc-direct/certhash/uEiAE_vuhjpPz5yq87PvAUAWca2yJxiuByzd_wRKiqcJGlw","/ip4/127.0.0.1/udp/4001/webrtc-direct/certhash/uEiAE_vuhjpPz5yq87PvAUAWca2yJxiuByzd_wRKiqcJGlw","/ip4/127.0.0.1/udp/4001/quic-v1","/ip6/2604:1380:45f1:d800::1/udp/4001/quic-v1","/ip6/2604:1380:45f1:d800::1/tcp/4001","/ip6/2604:1380:45f1:d800::1/tcp/4002/tls/sni/2604-1380-45f1-d800--1.k51qzi5uqu5dj0wvrbb8keygfyxe2v0fi1qbqz4pl3zzozle7oaqhf97mqazo4.libp2p.direct/ws","/ip4/145.40.89.101/udp/4001/quic-v1/webtransport/certhash/uEiBO8uH1QnET8rmwLLXlR5rocnu8gjZ1TZDQdMdp_yOowA/certhash/uEiA7PuBdt9syTOi4s3zSDE1XD2kBGN__66nSYaszeH3xRA","/ip6/::1/tcp/4001"],"ID":"12D3KooWHVXoJnv2ifmr9K6LWwJPXxkfvzZRHzjiTZMvybeTnwPy","Schema":"peer"}
{"ID":"12D3KooWJwuFqk7MkGDBbQ3Pz9EKQ3qn7m93SqZQFMNSq3bbp4Tq","Schema":"peer"}
{"ID":"12D3KooWLW7F5Kdrt21kPGKezLbJ2UTPuet7V6oQVJW1Z3VU1wdi","Schema":"peer"}
{"ID":"12D3KooWMbc9bbLBTFSmke99ZJETWTiSg3arCbMG6uZ7LMkX7WiG","Schema":"peer"}
{"ID":"12D3KooWMo7ZLNLxaCpGFHVJLEVDq6grrjULqvZ9eGuqY6DWuCAP","Schema":"peer"}
{"ID":"12D3KooWMqEqQiL5bsAE56Xg5nkWBW8o6cT6q7yvkjvCA5fnsAU2","Schema":"peer"}
{"ID":"12D3KooWPShtsFRTLtCJLpNVxjysSpmmzyy18rP4k7rh2TaDre5J","Schema":"peer"}
{"ID":"12D3KooWPajM7tGMpwvaDrBtVLsDmvoUMuqySm4tRTayowe1RJsQ","Schema":"peer"}
{"ID":"12D3KooWQCsg7PXQNQUGs2hJW7m1kXTr1Lt6MRgqwR8YazTcwEPP","Schema":"peer"}
{"ID":"12D3KooWQKKQCtxgQ3BNoTj5TgFxZywo7SimYK3Zyjr1K9EUnYes","Schema":"peer"}
{"Addrs":["/ip4/95.216.28.167/udp/4001/quic/p2p/12D3KooWPDBio8GH9svF7oS1mw849esxLs3gr5YViUGqzi2zwef9/p2p-circuit","/ip4/95.216.28.167/tcp/4001/p2p/12D3KooWPDBio8GH9svF7oS1mw849esxLs3gr5YViUGqzi2zwef9/p2p-circuit","/ip4/69.203.71.93/udp/64283/quic-v1/webtransport/certhash/uEiAZ3F5UJwcCIr3UPr_sPuWQj3v4HAasik_H_dLt0bzZTQ/certhash/uEiAPUcxZdKwDMsPCOs8qe4rmvLJRdNUVH6Mhi2nSy6EniQ/p2p/12D3KooWNMURc4AuvjC1vYARFTSKmMJBB1pjyr5zymBL15E96rqA/p2p-circuit","/ip4/69.203.71.93/tcp/64283/p2p/12D3KooWNMURc4AuvjC1vYARFTSKmMJBB1pjyr5zymBL15E96rqA/p2p-circuit","/ip4/69.203.71.93/udp/64283/quic-v1/p2p/12D3KooWNMURc4AuvjC1vYARFTSKmMJBB1pjyr5zymBL15E96rqA/p2p-circuit"],"ID":"12D3KooWSXS5KPLpPZ4jfCfek5SrCW2EeGFfkQr3fqb3UAikfX58","Schema":"peer"}
{"Addrs":["/ip4/66.29.147.177/tcp/4001/p2p/12D3KooWJNX7YDRBkh1m2vssumMHoP7PEn2gQ7WVs5s3vYxGfxxb/p2p-circuit","/ip4/66.29.147.177/udp/4001/quic-v1/webtransport/certhash/uEiBjZkayyaLxp88App6om86GYU2q-TB8I50gowq5MCCPqQ/certhash/uEiCwAXs3XknzDnLuXHkvwiJn-StgfExwnGPR-7pA5zsrmg/p2p/12D3KooWJNX7YDRBkh1m2vssumMHoP7PEn2gQ7WVs5s3vYxGfxxb/p2p-circuit","/ip4/66.29.147.177/udp/4001/quic-v1/p2p/12D3KooWJNX7YDRBkh1m2vssumMHoP7PEn2gQ7WVs5s3vYxGfxxb/p2p-circuit","/ip4/216.238.81.231/udp/4001/quic-v1/p2p/12D3KooWRo1uYUSm9wXn2LeBpZyQrU3KYeq2yzkKRSGc8FYjDXpb/p2p-circuit","/ip4/216.238.81.231/tcp/4001/p2p/12D3KooWRo1uYUSm9wXn2LeBpZyQrU3KYeq2yzkKRSGc8FYjDXpb/p2p-circuit","/ip4/216.238.81.231/udp/4001/quic/p2p/12D3KooWRo1uYUSm9wXn2LeBpZyQrU3KYeq2yzkKRSGc8FYjDXpb/p2p-circuit"],"ID":"12D3KooWEp8H3jDYw1Cn3hL432UX71Wv4UKmTXHP4mmL597GTmVn","Schema":"peer"}
```

2color replied:
I agree we should not return providers without addresses if we probed them.

At first this behaviour you caught here seemed was strange, but then I realised that this happens because we pass the context from the http request all the way to the additional `FindPeers`, which results in those `FindPeers` not completing due to an exceeded context deadline after the [default 30 seconds](https://github.com/ipfs/boxo/commit/e38f236348d62c9c75494c563975e3eccc745b10) which we recently added by default. 

A couple of options:
- Decouple the dispatched FindPeer from the request context. This is tricky to do, because the router here is an iterator rather than a channel, so it's a bit hard to handle this concurrently. 
- Change the default timeout
-

lidel commented on server_cached_router.go:
```diff
@@ -0,0 +1,230 @@
+			if !ok {
+				return false // channel closed. We're done
+			}
+			if result.Addrs != nil { // Only if the lookup returned a result and it has addrs
```
```suggestion
			if len(result.Addrs) > 0 { // Only if the lookup returned a result and it has addrs
```

lidel reviewed (changes requested):
I think this is mostly ready to ship / deploy early next week, but we need to address things marked with ⚠️:
- proactively remove peers that failed probing beyond max probing backoff
- skip peers that have no addrs and skipped probing

details and other smaller nits inline
(apologies if I misunderstood something, did this late Thursday, but wanted to give feedback before Friday)

2color committed:
Apply suggestions from code review

Co-authored-by: Marcin Rataj <lidel@lidel.org>

2color committed:
Apply suggestions from code review

Co-authored-by: Marcin Rataj <lidel@lidel.org>

2color committed:
fix: use separate context for dispatched jobs

2color committed:
fix: ensure proper cleanup of cache fallback iter

2color committed:
Update main.go

Co-authored-by: Marcin Rataj <lidel@lidel.org>

2color committed:
fix: formatting

2color committed:
fix: let consumer handle cleanup

2color committed:
fix: remove from address book when removed from peer state

2color committed:
fix: use normal lru cache instead of 2Q

2color committed:
fix: update the metric when removing from the peer cache

2color committed:
fix: increase max backoff to 48 hours

When the max backoff duration is reached and a connection attempt fails
we clear the cached addresses and state. Since this state is useful to
prevent unncessary attempts to dispatch a find peer we should keep it
for as long as a provider record is valid for.

2color commented:
> iiuc we don't have good mechanism for "forgetting" least peers, other than reaching PeerCacheSize limit here.
> 
> * we have a single cache for `peerState` success (online) and failures (offline)
> * over time the cache will be 1M of peers and most of them may be offline (due to regular churn, or network attacks - we've seen sudden spikes in random peers that then disappeared forever)
>   
>   * A good safeguard would be to explicitly remove peer from cache if failed for longer than MaxBackoffDuration (see comment below)
> * the cache will unlikely drop items often on its own  with `PeerCacheSize = 1_000_000`,  so there is no point in using 2Q, maybe switch to simpler LRU at this point


I made a couple of additions to your suggestion:
- When we remove the the `peerState` for a given peer, we also remove it from the cache. This is important, because otherwise, the peer will be probed again (since we deleted the state but we iterate on the address book for the probe)
- Increase the `MaxBackoffDuration` to 48 hours (from 24 hours). This is to delay when a peer is removed from address book and peerState cache. Why? Because even if a peer announces a provider record, but then goes offline, we want to know that their offline (through the peerState cache) for as long as that provider record is valid for, so that we don't dispatch FindPeer calls to that peer in the fallback router.

2color committed:
feat: add env var for recently connected ttl

2color committed:
feat: add env var to control active probing

2color commented on CHANGELOG.md:
```diff
@@ -15,6 +15,10 @@ The following emojis are used to highlight certain changes:
 
+- Default caching of peer addresses for 48h to match [provider record expiration on Amino DHT](https://github.com/libp2p/go-libp2p-kad-dht/blob/v0.28.1/amino/defaults.go#L40-L43). Someguy will return cached addresses for peers without multiaddrs in `FindProviders` if there are no addresses for a provider. This can be enabled via `SOMEGUY_CACHED_ADDR_BOOK=true|false` (enabled by default)
+  - Added a new `cachedAddrBook` implementation that caches peer addresses by subscribing to Identify events and probes those peers in the background.
+  - Added a new `cachedRouter` that uses `cachedAddrBook` to retrieve cached addresses for peers without multiaddrs. If a Peer is encountered with no cached addresses, `FindPeer` is dispatched in the background.
```
Added.

2color committed:
fix: bug from closing the iterator twice

no need to close the channel. just the source iterator

2color committed:
docs: update comment

2color committed:
docs: improve changelog

2color committed:
test: fix background test

lidel committed:
feat(metrics): track online vs offline probe ratio

lidel reviewed (approved):
Thank you @2color, lgtm, fyi added metric for tracking offline/online probe results (increase rate over time) – https://github.com/ipfs/someguy/pull/90/commits/c1ac41ba13a924a7d7c1b2676efe4c6d26bb8b70

Feel free to merge and release and deploy to https://delegated-ipfs.dev/ :pray: 
We will see if any follow-up is needed once we observe metrics in prod  (fine to pipe gateway cids to smoke-test there too).
 
---
 
Note for self / future reference: 
- The way the cache works in this PR is that we require at least one successful libp2p identify lookup for a peer to be accepted to the cache. 
  - Once in cache, we periodically probe to confirm if peer is online/offline, and refresh addrs 
  - If a cached peer turns offline, we note it in cache, and  stop returning it as one of providers, but we keep probing cached peers (incl. offline ones) with backoff just in case they are online again. If not, we drop once they been offline longer than the Amino DHT expiration window.
- Basing cache add on libp2p identify success means the cache does not include peerids that were returned by (dht?) router without addrs but "always failed to identify" 
  - I think it is ok to leave this as-is, if we ever need to decrease cost related to those, we can start caching these dead peers somewhere in "dispatched peer lookup" code to ensure they fall under backoff logic as well.

2color commented:
> Basing cache add on libp2p identify success means the cache does not include peerids that were returned by (dht?) router without addrs but "always failed to identify"

We only cache failure count and last failure since we don't have any addresses to cache for such peers.

> * I think it is ok to leave this as-is, if we ever need to decrease cost related to those, we can start caching these dead peers somewhere in "dispatched peer lookup" code to ensure they fall under backoff logic as well.

We already do.

When we dispatch a peer lookup, we use call `FindPeer` on the `cachedRouter`:
https://github.com/ipfs/someguy/blob/48e1943abb4f9b95d552ff057d3c05e156541e59/server_cached_router.go#L208

Which records failed connections:
https://github.com/ipfs/someguy/blob/48e1943abb4f9b95d552ff057d3c05e156541e59/server_cached_router.go#L77

Which ensures that the backoff logic applies to those peers too for up to MaxBackoffDuration (48 hours) of failures. If they continue to fail connecting for 48 hours they will be removed.

2color commented:
Another follow up improvement we could look into is checking the last failed connection in the `peerCache ` before blindly augmenting with addresses from the `cachedAddrBook`. 

As it currently stands, even if probing a peer fails, the cached addresses will still be returned, at least until `MaxBackoffDuration` of 48 hours is reached and the addresses are removed from cache.
